baseapp/streaming.go

```diff
@@ -18,6 +18,13 @@ type ABCIListener interface {
 	ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
 	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
 	ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
+	// HaltAppOnDeliveryError returns true if the application has been configured to halt when
+	// ListenBeginBlock, ListenEndBlock, ListenDeliverTx fail to process messages and false when
+	// the application has been configured to send messages to ListenBeginBlock, ListenEndBlock, ListenDeliverTx
+	// in fire-and-forget fashion.
+	//
+	// This behavior is controlled by a corresponding app config setting.
+	HaltAppOnDeliveryError() bool
```

Comment:
`app.halt()` is private so it would not be possible to halt the app from the listeners. Also, do we want to expose more of the internal API to the listeners? https://github.com/cosmos/cosmos-sdk/blob/4385325cc13b6263232ec28276fa631be0332a06/baseapp/abci.go#L355

@i-norden ^^^

Reply:
It seems that `app.halt` just kills the process un-cleanly. Presuming that is the intended behavior, you don't need a special method to do that :) You can just `os.Exit(0)` anywhere.

Reply:
:) very true but it's an unclean way of exiting. Also, it will bypass error reporting in the listener methods. `ListenBeginBlock, ListenEndBlock, ListenDeliverTx` report errors up the stack so their method signature will need to change. 

We should have the the broader team chime in on this a bit and provide feedback as to how we want plugins to behave in regards to this.

`app.halt` does eventually `os.Exit(0)` so perhaps we just fix its behavior?

Reply:
@egaxhaj 

> very true but it's an unclean way of exiting. Also, it will bypass error reporting in the listener methods . . . 

Sure! Halting an app is (AFAIK) an unclean exit, by definition. I don't think there's any way to reliably issue a "halt" from within the SDK that (a) actually terminates the process, _and_ (b) unwinds the call stack, and captures/logs errors, in a clean and predictable way. Happy to be corrected.

> app.halt does eventually os.Exit(0) so perhaps we just fix its behavior?

Is the current behavior not correct? What would be more correct?

Reply:
my 2 cents is this is a secondary system that should have no effect on the app and consensus. If the indexer is halting the app to not lose data then a different approach needs to be thought of in which it doesn't affect the app.

Reply:
@marbar3778 - @i-norden and I have had [discussions](https://github.com/cosmos/cosmos-sdk/pull/10482#issuecomment-975840756) about how a secondary system would work that does not effect the app. However, we agreed to move forward with the current design to allow for time to be able to work through a secondary system design.